pydantic/networks.py

```diff
@@ -95,17 +95,176 @@ def defined_constraints(self) -> dict[str, Any]:
         return {field.name: getattr(self, field.name) for field in fields(self)}
 
 
-# TODO: there's a lot of repeated code in these two base classes - should we consolidate, or does that up
-# the complexity enough that it's not worth saving a few lines?
+class _BaseUrl:
+    _constraints: ClassVar[UrlConstraints] = UrlConstraints()
+    _url: Url
 
+    def __init__(self, url: str | Url) -> None:
+        if isinstance(url, Url):
```

Comment:
this is wrong, constraints won't be maintained.

I think you want:

```py
if isinstance(url, self.__class__):
    self._url = url._url
elif isinstance(url, (_BaseUrl, Url, _BaseMultiHostUrl, MultiHostUrl):
    url = str(url)
    ... do validation
else:
    do valdiation
```

Reply:
I think the point is that this these types are already broken in the same way in the latest release, I was happy to see constraint validation fixed in a future follow up.

Reply:
I think it's not too hard to just get it right now.

Reply:
Yeah. Happy to fix this now, but indeed, the previous approach was assuming we'd fix the init constraint behavior in a future PR.